sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -529,6 +530,15 @@ object SimplifyBinaryComparison
         case TrueLiteral EqualNullSafe b if !b.nullable => b
         case a EqualNullSafe FalseLiteral if !a.nullable => Not(a)
         case FalseLiteral EqualNullSafe b if !b.nullable => Not(b)
+
+        case EqualNullSafe(a: ScalarSubquery, b: ScalarSubquery)
+            if a.plan.maxRows.contains(0) && b.plan.maxRows.contains(0) =>
+          TrueLiteral
+        case e: EqualNullSafe => e
+        case BinaryComparison(a: ScalarSubquery, _) if a.plan.maxRows.contains(0) =>
+          Literal(null, BooleanType)
```

Comment:
Although all the current sub class of `BinaryComparison` are also with NullIntolerant.
In order to doesn't introduce unexpected bug, shall we check `isInstanceOf[NullIntolerant]`?

Reply:
Should move it to `ConstantFolding `.